### PR TITLE
 blog: Post about how to achieve baremetal performance

### PR DESCRIPTION
**Pull request by sboeuf**
This post details the reason why using VFIO or vDPA can help a user reach native performance of a device inside a virtual machine.

### CODE REVIEW

File: content/en/blog/baremetal_performance/index.md [hunk 1]
```diff
@@ -0,0 +1,152 @@
+---
+title: "How to achieve baremetal performance inside your virtual machine"
```
Comment (rbradford) [inline content/en/blog/baremetal_performance/index.md:2]: Could you shorten this to "Achieving baremetal performance within a Virtual Machine"

Comment (rbradford) [reply content/en/blog/baremetal_performance/index.md:2]: I don't know if baremetal is one word, two or hyphenated. Would be good to check style guides (e.g. Google's)

File: content/en/blog/baremetal_performance/index.md [hunk 2]
```diff
@@ -0,0 +1,152 @@
+---
+title: "How to achieve baremetal performance inside your virtual machine"
+date: 2022-05-19T08:00:00Z
+draft: false
+contributors: [Cloud Hypervisor Team]
+---
+
+There are use cases for which a workload needs to access specific hardware such
+as accelerators, GPU or network adapters to reach high performance. And since
```
Comment (rbradford) [inline content/en/blog/baremetal_performance/index.md:9]: Maybe "to reach high peformance" could be "to maximise potential performance"

File: content/en/blog/baremetal_performance/index.md [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+contributors: [Cloud Hypervisor Team]
+---
+
+There are use cases for which a workload needs to access specific hardware such
+as accelerators, GPU or network adapters to reach high performance. And since
+these workloads run inside virtual machines (VM) for security reasons, the
+challenge is to make this hardware available from within the VM, but without
+degrading the performance that can be achieved from baremetal.
+
+### VFIO, the first choice
```
Comment (rbradford) [inline content/en/blog/baremetal_performance/index.md:14]: Since you're using title case for the headings it might be best if you're consistent.

Comment (rbradford) [reply content/en/blog/baremetal_performance/index.md:14]: By first do you mean ideal or 1st vs 2nd? If the former maybe it's best to use that word instead.

File: content/en/blog/baremetal_performance/index.md [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+
+This is a framework that has been around for quite some time now, allowing a
```
Comment (rbradford) [inline content/en/blog/baremetal_performance/index.md:16]: Maybe you could use the word "mature" to simplify this sentence.

File: content/en/blog/baremetal_performance/index.md [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+
+It exposes an ioctl interface so that a userspace program can interact with the
+device and retrieve all the information it needs. In the virtualization case,
+this userspace program is the Virtual Machine Monitor (VMM), Cloud Hypervisor
+for instance, which after it got all the information from the device, exposes
+them to the guest. And that is how the workload can access the device inside the
+VM, exactly the same way it would if it was directly running on baremetal.
+
+Now that we've covered the functional aspect, let's have a look at the way we
+can achieve native performance from the guest.
```
Comment (rbradford) [inline content/en/blog/baremetal_performance/index.md:28]: You're using native here instead of baremetal. Would be best to stick with one piece of terminology.

File: content/en/blog/baremetal_performance/index.md [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+
+#### How to use VFIO
+
+For more details on how to use VFIO with Cloud Hypervisor, refer to the
+following [documentation](https://github.com/cloud-hypervisor/cloud-hypervisor/blob/main/docs/vfio.md).
+
+### vDPA, an interesting alternative
+
+vDPA stands for virtio Data Path Acceleration, and the main reason for this new
+framework for being created is its ability to simplify the migration process.
```
Comment (rbradford) [inline content/en/blog/baremetal_performance/index.md:102]: I think the second "for" in this sentence is unnecessary

File: content/en/blog/baremetal_performance/index.md [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+The whole concept behind vDPA relies on the virtio specification, which is what
+makes it more suited for migration.
+
+#### Control path
+
+On one hand, the virtio control path is used for configuring the virtqueues and
```
Comment (rbradford) [inline content/en/blog/baremetal_performance/index.md:108]: Do you mean "On the one hand" ?

File: content/en/blog/baremetal_performance/index.md [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+host. But in this very special case, the virtqueues are directly part of the
+physical device, allowing DMA transfers to be initiated by the guest. And the
+same way it works with VFIO because the device is also attached to a physical
+IOMMU, the guest can perform secure and efficient memory accesses from the
+device.
+
+#### Migration (experimental)
+
+Providing a simpler path for migrating physical devices from one VM to another
+is one of the main reason vDPA was created for. Because of the design relying
```
Comment (rbradford) [inline content/en/blog/baremetal_performance/index.md:128]: I don't think you need the final "for"

**Review by rbradford — APPROVED** [review]
This is good. I don't want to interfere too much with the style as I would like to preserve that it is written in your style (not mine)

File: content/en/blog/baremetal_performance/index.md (hunk 3 above)
Comment (sboeuf) [reply content/en/blog/baremetal_performance/index.md:14]: Ideal is better.

**Comment by sboeuf** [comment]
Thanks for the review @rbradford, comments have been addressed.

**Comment by rbradford** [comment]
@sboeuf Merge when you're ready
